src/riak_cs_bw_limitter.erl

```diff
@@ -0,0 +1,235 @@
+        end,
+    case test_bandwidth_state(IntervalSec, UserState) of
+        ok ->
+            case test_access_state(IntervalSec, UserState) of
+                ok ->
+                    {ok, RD, Ctx};
+                {error, {_, Current, Threshold}} = Error ->
+                    lager:warning("User ~p has exceeded access limit: usage, limit = ~p, ~p (/sec)",
+                                  [OwnerKey, Current, Threshold]),
+                    Error
+            end;
+        {error, {_, Current, Threshold}} = Error2 ->
+            lager:warning("User ~p has exceeded bandwidth limit: usage, limit = ~p, ~p (bytes/sec)",
+                          [OwnerKey, Current, Threshold]),
+            Error2
+    end.
+
+-spec new_user_state(binary()) -> #user_state{}.
+new_user_state(User) ->
+    UserState = #user_state{user = User},
+    lager:debug("quota init: ~p => ~p", [User, UserState]),
+    %% Here's a race condition where if so many concurrent access
+    %% come, each access can yield a new fresh state and then
+    %% receives not access limitation.
+    ets:insert_new(?MODULE, UserState),
+    UserState.
+
+test_bandwidth_state(DurationSec,
+                     #user_state{bandwidth = Bandwidth}) ->
+    BandwidthMax =
+        case application:get_env(riak_cs, bw_limitter_bandwidth_limit) of
+            {ok, V} when is_integer(V) -> V;
+            _ -> -1
+        end,
+    case Bandwidth / DurationSec < BandwidthMax  of
+        true -> ok;
+        false when BandwidthMax < 0 -> ok; %% No quota set
+        false -> {error, {bandwidth, Bandwidth, BandwidthMax}}
+    end.
+
+test_access_state(IntervalSec,
+                  #user_state{access_count = AccessCount}) ->
+    AccessCountMax =
+        case application:get_env(riak_cs, bw_limitter_rate_limit) of
+            {ok, V} when is_integer(V) -> V;
+            _ -> -1
+        end,
+    case AccessCount / IntervalSec < AccessCountMax of
+        true -> ok;
+        false when AccessCountMax < 0 -> ok; %% No quota set
+        false -> {error, {access_count, AccessCount, AccessCountMax}}
+    end.
+
+-spec update(binary(), wm_log_data()) -> ok | {error, term()}.
+update(User,
+       #wm_log_data{response_length = ResponseLength,
+                    headers = Headers} = _LogData) ->
+    Bytes0 = case is_integer(ResponseLength) of
+                true -> ResponseLength;
+                _ -> 0
+            end,
+    RequestLength = mochiweb_headers:get_value("content-length", Headers),
+    Bytes = case catch list_to_integer(RequestLength) of
+                Len when is_integer(Len) -> Len + Bytes0;
+                _ -> Bytes0
+            end,
+    UpdateOps = [{#user_state.bandwidth, Bytes},
+                 {#user_state.access_count, 1}],
+    try
+        case ets:update_counter(?MODULE, User, UpdateOps) of
+            [_, _] = R ->
+                ok;
+            Error0 ->
+                lager:warning("something wrong? ~p", [Error0]),
```

Comment:
Depending on refresh interval, this warning logs can be spammy. Race example steps:

1. get user_state record from ets
2. refresh triggered
3. ets truncated
4. update_counter